generator/transit_generator.cpp

```diff
@@ -49,121 +48,92 @@ using namespace std;
+  }
+  void operator()(vector<transit::Edge> & c, char const * /* name */ = nullptr) const {}
+};
 
-void DeserializeGatesFromJson(my::Json const & root, string const & mwmDir, string const & countryId,
-                              OsmIdToFeatureIdsMap const & osmIdToFeatureIdsMap, vector<Gate> & gates)
+struct IsValidVisitor
 {
-  DeserializeFromJson(root, "gates", osmIdToFeatureIdsMap, gates);
-
-  // Creating IndexRouter.
-  SingleMwmIndex index(my::JoinFoldersToPath(mwmDir, countryId + DATA_FILE_EXTENSION));
-
-  auto infoGetter = storage::CountryInfoReader::CreateCountryInfoReader(GetPlatform());
-  CHECK(infoGetter, ());
-
-  auto const countryFileGetter = [&infoGetter](m2::PointD const & pt) {
-    return infoGetter->GetRegionCountryId(pt);
-  };
+  template <typename Cont>
+  void operator()(Cont const & c, char const * /* name */ = nullptr)
+  {
+    if (!m_isValid)
+      return;
 
-  auto const getMwmRectByName = [&](string const & c) -> m2::RectD {
-    CHECK_EQUAL(countryId, c, ());
-    return infoGetter->GetLimitRectForLeaf(c);
-  };
+    m_isValid = ::IsValid(c);
+  }
 
-  CHECK_EQUAL(index.GetMwmId().GetInfo()->GetType(), MwmInfo::COUNTRY, ());
-  auto numMwmIds = make_shared<NumMwmIds>();
-  numMwmIds->RegisterFile(CountryFile(countryId));
+  bool IsValid() const { return m_isValid; }
 
-  // Note. |indexRouter| is valid while |index| is valid.
-  IndexRouter indexRouter(VehicleType::Pedestrian, false /* load altitudes */,
-                          CountryParentNameGetterFn(), countryFileGetter, getMwmRectByName,
-                          numMwmIds, MakeNumMwmTree(*numMwmIds, *infoGetter),
-                          traffic::TrafficCache(), index.GetIndex());
+private:
+  bool m_isValid = true;
+};
 
-  // Looking for the best segment for every gate.
-  for (auto & gate : gates)
+struct IsUniqueVisitor
+{
+  template <typename Cont>
+  void operator()(Cont const & c, char const * /* name */ = nullptr)
   {
-    // Note. For pedestrian routing all the segments are considered as twoway segments so
-    // IndexRouter.FindBestSegment() method finds the same segment for |isOutgoing| == true
-    // and |isOutgoing| == false.
-    Segment bestSegment;
-    try
-    {
-      if (indexRouter.FindBestSegmentInSingleMwm(gate.GetPoint(),
-                                                 m2::PointD::Zero() /* direction */,
-                                                 true /* isOutgoing */, bestSegment))
-      {
-        CHECK_EQUAL(bestSegment.GetMwmId(), 0, ());
-        gate.SetBestPedestrianSegment(SingleMwmSegment(
-            bestSegment.GetFeatureId(), bestSegment.GetSegmentIdx(), bestSegment.IsForward()));
-      }
-    }
-    catch (RootException const & e)
-    {
-      LOG(LDEBUG, ("Point of a gate belongs to several mwms or doesn't belong to any mwm. Gate:",
-                   gate, e.what()));
-    }
+    if (!m_isUnique)
+      return;
+
+    m_isUnique = (adjacent_find(c.begin(), c.end()) == c.end());
```

Comment:
Will be it useful to, at least, emit warn messages for containers that contain duplicates? In this case, it may be useful to get rid of early exit.

Reply:
A duplicate mean a significant error which is enough to stop generator. So if |m_isUnique| == false GraphData::IsValid() returns false and generator will be terminate in ProcessGraph() (called from BuildTransit())

Probably it's worth renaming methods of GraphData to IsValid(), IsUnique() and IsSorted() to CheckValid(), CheckUnique() and CheckSorted() and place CHECK inside them. 

What do you think about it?